janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/StandardJanusGraphTx.java

```diff
@@ -1252,49 +1253,71 @@ public boolean isDeleted(VertexCentricQuery query, JanusGraphRelation result) {
             if (query.numSubQueries()==1 && query.getSubQuery(0).getBackendQuery().isEmpty())
                 return Collections.emptyIterator();
             Preconditions.checkArgument(query.getCondition().hasChildren(),"If the query is non-empty it needs to have a condition");
-
             if (query.getResultType() == ElementCategory.VERTEX && hasModifications()) {
                 Preconditions.checkArgument(QueryUtil.isQueryNormalForm(query.getCondition()));
-                PredicateCondition<PropertyKey, JanusGraphElement> standardIndexKey = getEqualityCondition(query.getCondition());
+                List<PredicateCondition<PropertyKey, JanusGraphElement>> equalityConditions = getEqualityConditions(query.getCondition());
                 Iterator<JanusGraphVertex> vertices;
-                if (standardIndexKey == null) {
-                    final Set<PropertyKey> keys = Sets.newHashSet();
-                    ConditionUtil.traversal(query.getCondition(), cond -> {
-                        Preconditions.checkArgument(cond.getType() != Condition.Type.LITERAL || cond instanceof PredicateCondition);
-                        if (cond instanceof PredicateCondition) {
-                            keys.add(((PredicateCondition<PropertyKey, JanusGraphElement>) cond).getKey());
+                if (equalityConditions.isEmpty()) {
+                    vertices = withoutCache(query);
+                } else {
+                    int minSize = Integer.MAX_VALUE;
+                    PredicateCondition<PropertyKey, JanusGraphElement> bestCond = null;
+                    for (PredicateCondition<PropertyKey, JanusGraphElement> cond : equalityConditions) {
+                        int currentSize = Iterables.size(newVertexIndexEntries.get(cond.getValue(), cond.getKey()));
```

Comment:
Part of the decreased performance of QUERY#2 could be due to the call of `newVertexIndexEntries.get(...)` here.
Looking at the [current implementation](https://github.com/JanusGraph/janusgraph/blob/master/janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/indexcache/ConcurrentIndexCache.java) this creates an `ArrayList` and populates it with the matching index entries. Although `Iterables` is clever enough to detect that the resulting `Iterable` is in fact a `Collection` and calls it's `size()` method directly, this looks to me like an overkill for counting the number of results.

Instead, it could be beneficial to extend the Interface `IndexCache` by a method `count(Object value, PropertyKey key)` which only returns the size of the result set directly.